drivers/gpio/gpio_aw9523b.c

```diff
@@ -0,0 +1,474 @@
+/*
+ * Copyright (c) 2024 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT awinic_aw9523b_gpio
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/mfd/aw9523b.h>
+#include <zephyr/sys/ring_buffer.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(gpio_aw9523b, CONFIG_GPIO_LOG_LEVEL);
+
+#define AW9523B_GPOMD             BIT(4)
+#define AW9523B_RESET_PULSE_WIDTH 20
+
+enum read_write_toggle_t {
+	READ,
+	WRITE,
+	TOGGLE,
+};
+
+struct gpio_aw9523b_config {
+	struct gpio_driver_config common;
+	const struct device *mfd_dev;
+	struct i2c_dt_spec i2c;
+	bool port0_push_pull;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+	struct gpio_dt_spec reset_gpio;
+#endif
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	struct gpio_dt_spec int_gpio;
+	gpio_callback_handler_t int_cb;
+#endif
+};
+
+struct gpio_aw9523b_data {
+	struct gpio_driver_data common;
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	const struct device *dev;
+	sys_slist_t callbacks;
+	struct gpio_callback gpio_callback;
+	struct k_work intr_worker;
+	gpio_port_value_t prev_value;
+	gpio_port_pins_t rising_event_pins;
+	gpio_port_pins_t falling_event_pins;
+#endif
+};
+
+static int gpio_aw9523b_pin_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_aw9523b_config *const config = dev->config;
+	const uint8_t port = (pin < 8) ? 0 : 1;
+	const uint8_t mask = BIT(pin % 8);
+	const uint8_t input_en = (flags & GPIO_INPUT) ? mask : 0x00;
+	const uint8_t out_high = (flags & GPIO_OUTPUT_INIT_HIGH) ? mask : 0x00;
+	int err;
+
+	/* Can't do I2C operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Either INPUT or OUTPUT must be set */
+	if ((!!(flags & GPIO_INPUT) ^ !!(flags & GPIO_OUTPUT)) == 0) {
```

Comment:
that's a funny way of doing it, how about

`if ((flags & GPIO_INPUT) != 0 && (flags & GPIO_OUTPUT) != 0)`? bit easier to follow

Reply:
Here, I do XOR operations. The equivariant equation is $`{\displaystyle P\veebar Q=(P\lor Q)\land (\lnot P\lor \lnot Q)}`$
I corrected these lines following the equation.